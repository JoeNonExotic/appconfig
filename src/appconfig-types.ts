/**
 * A simple type represenation for https://www.rfc-editor.org/rfc/rfc9110.html#name-content-type "application-json"
 */
export type ConfigurationContent = { [x: string]: any } | [{ [x: string]: any }];

/**
 * A simple type representation for https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-appconfig-configurationprofile.html#cfn-appconfig-configurationprofile-type
 */
export type ConfigurationProfileType = 'AWS.Freeform' | 'AWS.AppConfig.FeatureFlags';

/**
 * A simple type representation for https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-appconfig-deploymentstrategy.html#cfn-appconfig-deploymentstrategy-replicateto
 */
export type DeploymentStrategyReplication = 'NONE' | 'SSM_DOCUMENT';