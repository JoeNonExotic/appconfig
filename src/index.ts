export * from './simple-appconfig';
export * from './appconfig-types';
export * from './utils';