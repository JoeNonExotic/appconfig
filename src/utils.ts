import { CfnApplication, CfnConfigurationProfile, CfnDeployment, CfnDeploymentStrategy, CfnEnvironment, CfnHostedConfigurationVersion } from 'aws-cdk-lib/aws-appconfig';
import { Construct } from 'constructs';
import { ConfigurationContent, ConfigurationProfileType, DeploymentStrategyReplication } from './appconfig-types';

/**
 * Gets a new {@link CfnApplication} with the specified params.
 * @param scope The scope of the construct.
 * @param id A unique id.
 * @param applicationName The name of the application.
 * @returns an initialized {@link CfnApplication}.
 */
export function getApplication(
  scope: Construct,
  id: string,
  applicationName: string,
): CfnApplication {
  return new CfnApplication(
    scope,
    id,
    {
      name: applicationName,
    },
  );
}

/**
 * Gets a new {@link CfnDeploymentStrategy} with the specified params.
 * @param scope The scope of the construct.
 * @param id A unique id.
 * @param {string} [deploymentName=id] An optional deploymentName, defaults to id if not specified.
 * @param {number} [deploymentDurationInMinutes=0] An optional deployment time in minutes, defaults to 0 if not specified.
 * @param {number} [growthFactor=100] An optional growth factor, defaults to 100 if not specified.
 * @param {string} [replicateTo='NONE'] An optional replicateTo, defaults to 'NONE' if not specified.
 * @param {number} [finalBakeTimeInMinutes=0] An optional final bake time, defaults to 0 if not specified.
 * @returns an initialized {@link CfnDeploymentStrategy}.
 */
export function getAppConfigDeploymentStrategy(
  scope: Construct,
  id: string,
  deploymentName?: string,
  deploymentDurationInMinutes?: number,
  growthFactor?: number,
  replicateTo?: DeploymentStrategyReplication,
  finalBakeTimeInMinutes?: number,
): CfnDeploymentStrategy {
  return new CfnDeploymentStrategy(scope, id, {
    name: deploymentName ? deploymentName : id,
    deploymentDurationInMinutes: deploymentDurationInMinutes ? deploymentDurationInMinutes : 0,
    growthFactor: growthFactor ? growthFactor : 100,
    replicateTo: replicateTo ? replicateTo : 'NONE',
    finalBakeTimeInMinutes: finalBakeTimeInMinutes ? finalBakeTimeInMinutes : 0,
  });
}

/**
 * Gets a new {@link CfnDeploymentStrategy} with the specified params.
 * @param scope The scope of the construct.
 * @param id A unique id.
 * @param application A valid {@link CfnApplication}.
 * @param envName Name for the app config environment.
 * @returns an initialized {@link CfnEnvironment}.
 */
export function getAppConfigEnvironment(
  scope: Construct,
  id: string,
  application: CfnApplication,
  envName: string,
): CfnEnvironment {
  return new CfnEnvironment(scope, id, {
    applicationId: application.ref,
    // can be anything that makes sense for your use case.
    name: envName,
  });
}

/**
 * Gets a new {@link CfnConfigurationProfile} with the specified params.
 * @param scope The scope of the construct.
 * @param id A unique id.
 * @param application A valid {@link CfnApplication}.
 * @param envName Name for the app config environment.
 * @returns an initialized {@link CfnConfigurationProfile}.
 */
export function getAppConfigHostedConfigurationProfile(
  scope: Construct,
  id: string,
  application: CfnApplication,
  configurationProfileName: string,
  configurationProfileType: ConfigurationProfileType,
): CfnConfigurationProfile {
  return new CfnConfigurationProfile(
    scope,
    id,
    {
      name: configurationProfileName,
      applicationId: application.ref,
      locationUri: 'hosted',
      type: configurationProfileType,
    },
  );
}

/**
 * Gets a new {@link CfnHostedConfigurationVersion} with the specified params.
 * @param scope The scope of the construct.
 * @param id A unique id.
 * @param application A valid {@link CfnApplication}.
 * @param configurationProfile A valid {@link CfnConfigurationProfile}.
 * @param configurationConent Content matching {@link ConfigurationContent}.
 * @returns an initialized {@link CfnConfigurationProfile}
 */
export function getAppConfigHostedConfigurationVersion(
  scope: Construct,
  id: string,
  application: CfnApplication,
  configurationProfile: CfnConfigurationProfile,
  configurationConent: ConfigurationContent): CfnHostedConfigurationVersion {
  return new CfnHostedConfigurationVersion(
    scope,
    id,
    {
      applicationId: application.ref,
      configurationProfileId: configurationProfile.ref,
      content: JSON.stringify(configurationConent),
      contentType: 'application/json',
    },
  );
}

/**
 * Gets a new {@link CfnDeployment} with the specified params.
 * @param scope The scope of the construct.
 * @param id A unique id.
 * @param application A valid {@link CfnApplication}.
 * @param configurationProfile A valid {@link CfnConfigurationProfile}.
 * @param configurationVersion A valid {@link CfnHostedConfigurationVersion}.
 * @param deploymentStrategy A valid {@link CfnDeploymentStrategy}.
 * @param environment A valid {@link CfnEnvironment}.
 * @returns an initialized {@link CfnDeployment}.
 */
export function getAppConfigDeployment(
  scope: Construct,
  id: string,
  application: CfnApplication,
  configurationProfile: CfnConfigurationProfile,
  configurationVersion: CfnHostedConfigurationVersion,
  deploymentStrategy: CfnDeploymentStrategy,
  environment: CfnEnvironment): CfnDeployment {
  return new CfnDeployment(scope, id, {
    applicationId: application.ref,
    configurationProfileId: configurationProfile.ref,
    configurationVersion: configurationVersion.ref,
    deploymentStrategyId: deploymentStrategy.ref,
    environmentId: environment.ref,
  });
}
