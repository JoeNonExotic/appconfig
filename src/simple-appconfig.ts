import { CfnApplication, CfnConfigurationProfile, CfnDeployment, CfnDeploymentStrategy, CfnEnvironment, CfnHostedConfigurationVersion } from 'aws-cdk-lib/aws-appconfig';
import { Construct } from 'constructs';
import { ConfigurationContent } from './appconfig-types';
import { getApplication, getAppConfigDeploymentStrategy, getAppConfigEnvironment, getAppConfigHostedConfigurationProfile, getAppConfigHostedConfigurationVersion, getAppConfigDeployment } from './utils';

/**
 * Props for {@link SimpleAppConfig} construct.
 */
export interface SimpleAppConfigProps {
  /**
     * The name of the app config application. (Ex: MyApplication)
     */
  readonly applicationName: string;

  /**
     * The name of the environment for deployment. (Ex: Production)
     */
  readonly applicationEnvironment: string;

  /**
     * The name for the configuration profile within the application.
     */
  readonly configurationProfileName: string;

  /**
     * The content within the configuration profile.
     */
  readonly configurationContent: ConfigurationContent;
}

/**
 * Custom construct to quickly setup app config resouce based on the passed in {@link SimpleAppConfigProps}.
 * This construct does the following:
 * 1. Setup `applicationName/applicationEnvironment/configurationProfileName`.
 * 2. Add/update configurationContent.
 * 3. Apply the add/update immediately.
 */
export class SimpleAppConfig extends Construct {

  public readonly application: CfnApplication;
  public readonly immediateDeploymentStrategy: CfnDeploymentStrategy;
  public readonly environment: CfnEnvironment;
  public readonly configurationProfile: CfnConfigurationProfile;
  public readonly hostedConfigurationVersion: CfnHostedConfigurationVersion;
  public readonly deployment: CfnDeployment;

  constructor(
    scope: Construct,
    id: string,
    private readonly props: SimpleAppConfigProps,
  ) {
    super(scope, id);

    // Create a new application
    this.application = getApplication(
      this,
      'Application',
      this.props.applicationName,
    );


    // setup a deploy strategy to immediately apply config.
    this.immediateDeploymentStrategy = getAppConfigDeploymentStrategy(
      this,
      'ImmediateDeployStrategy',
      'ImmediateDeployment',
    );

    // setup an app config env
    this.environment = getAppConfigEnvironment(
      this,
      'Environment',
      this.application,
      this.props.applicationEnvironment,
    );

    // setup config profile
    this.configurationProfile = getAppConfigHostedConfigurationProfile(
      this,
      'ConfigurationProfile',
      this.application,
      this.props.configurationProfileName,
      'AWS.Freeform',
    );

    // Update AppConfig
    this.hostedConfigurationVersion = getAppConfigHostedConfigurationVersion(
      this,
      'HostedConfigurationVersion',
      this.application,
      this.configurationProfile,
      this.props.configurationContent,
    );


    // Perform deployment.
    this.deployment = getAppConfigDeployment(
      this,
      'Deployment',
      this.application,
      this.configurationProfile,
      this.hostedConfigurationVersion,
      this.immediateDeploymentStrategy,
      this.environment,
    );
  }
}