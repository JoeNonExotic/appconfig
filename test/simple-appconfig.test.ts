import { App, assertions, Stack } from 'aws-cdk-lib';
import { SimpleAppConfig } from '../src';

/**
 * Test configuration content.
 */
const TEST_CONFIGURATION = {
  StringKey: 'StringValue',
  BooleanKey: true,
  NumberKey: 123,
  CompositeKey: {
    StringKey: 'StringValue',
    BooleanKey: false,
    NumberKey: 456,
  },
  ArrayKey: [
    {
      StringKey: 'StringValue1',
      BooleanKey: true,
      NumberKey: 123,
    },
    {
      StringKey: 'StringValue2',
      BooleanKey: false,
      NumberKey: 456,
    },
  ],
};

const TEST_APP_NAME = 'MyTestApp';
const TEST_ENV_NAME = 'MyTestEnvironment';
const TEST_CONFIG_PROFILE_NAME = 'MyConfigurationProfile';

describe('simple-appconfig', () => {
  it('creates expected resources', () => {
    const app = new App();
    const stack = new Stack(app, 'MyTestStack');
    new SimpleAppConfig(stack, 'SimpleAppConfig', {
      applicationName: TEST_APP_NAME,
      applicationEnvironment: TEST_ENV_NAME,
      configurationProfileName: TEST_CONFIG_PROFILE_NAME,
      configurationContent: TEST_CONFIGURATION,
    });

    const template = assertions.Template.fromStack(stack);
    expect(template).toMatchSnapshot();

    template.hasResourceProperties('AWS::AppConfig::Application', {
      Name: TEST_APP_NAME,
    });
    template.resourceCountIs('AWS::AppConfig::Application', 1);

    template.hasResourceProperties('AWS::AppConfig::ConfigurationProfile', {
      ApplicationId: { Ref: 'SimpleAppConfigApplication04939DED' },
      LocationUri: 'hosted',
      Name: TEST_CONFIG_PROFILE_NAME,
      Type: 'AWS.Freeform',
    });
    template.resourceCountIs('AWS::AppConfig::ConfigurationProfile', 1);


    template.hasResourceProperties('AWS::AppConfig::DeploymentStrategy', {
      DeploymentDurationInMinutes: 0,
      FinalBakeTimeInMinutes: 0,
      GrowthFactor: 100,
      Name: 'ImmediateDeployment',
      ReplicateTo: 'NONE',
    });
    template.resourceCountIs('AWS::AppConfig::DeploymentStrategy', 1);

    template.hasResourceProperties('AWS::AppConfig::Deployment', {
      ApplicationId: {
        Ref: 'SimpleAppConfigApplication04939DED',
      },
      ConfigurationProfileId: {
        Ref: 'SimpleAppConfigConfigurationProfile78536DC3',
      },
      ConfigurationVersion: {
        Ref: 'SimpleAppConfigHostedConfigurationVersion67E8BB89',
      },
      DeploymentStrategyId: {
        Ref: 'SimpleAppConfigImmediateDeployStrategy781EA584',
      },
      EnvironmentId: {
        Ref: 'SimpleAppConfigEnvironmentCD8BFB59',
      },
    });
    template.resourceCountIs('AWS::AppConfig::Deployment', 1);

    template.hasResourceProperties('AWS::AppConfig::Environment', {
      ApplicationId: {
        Ref: 'SimpleAppConfigApplication04939DED',
      },
      Name: TEST_ENV_NAME,
    });
    template.resourceCountIs('AWS::AppConfig::Environment', 1);

    template.hasResourceProperties('AWS::AppConfig::HostedConfigurationVersion', {
      ApplicationId: {
        Ref: 'SimpleAppConfigApplication04939DED',
      },
      ConfigurationProfileId: {
        Ref: 'SimpleAppConfigConfigurationProfile78536DC3',
      },
      Content: JSON.stringify(TEST_CONFIGURATION),
      ContentType: 'application/json',
    });
    template.resourceCountIs('AWS::AppConfig::HostedConfigurationVersion', 1);
  });
});