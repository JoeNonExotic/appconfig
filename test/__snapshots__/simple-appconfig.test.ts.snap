// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-appconfig creates expected resources 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "SimpleAppConfigApplication04939DED": Object {
      "Properties": Object {
        "Name": "MyTestApp",
      },
      "Type": "AWS::AppConfig::Application",
    },
    "SimpleAppConfigConfigurationProfile78536DC3": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "SimpleAppConfigApplication04939DED",
        },
        "LocationUri": "hosted",
        "Name": "MyConfigurationProfile",
        "Type": "AWS.Freeform",
      },
      "Type": "AWS::AppConfig::ConfigurationProfile",
    },
    "SimpleAppConfigDeploymentB0181BC9": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "SimpleAppConfigApplication04939DED",
        },
        "ConfigurationProfileId": Object {
          "Ref": "SimpleAppConfigConfigurationProfile78536DC3",
        },
        "ConfigurationVersion": Object {
          "Ref": "SimpleAppConfigHostedConfigurationVersion67E8BB89",
        },
        "DeploymentStrategyId": Object {
          "Ref": "SimpleAppConfigImmediateDeployStrategy781EA584",
        },
        "EnvironmentId": Object {
          "Ref": "SimpleAppConfigEnvironmentCD8BFB59",
        },
      },
      "Type": "AWS::AppConfig::Deployment",
    },
    "SimpleAppConfigEnvironmentCD8BFB59": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "SimpleAppConfigApplication04939DED",
        },
        "Name": "MyTestEnvironment",
      },
      "Type": "AWS::AppConfig::Environment",
    },
    "SimpleAppConfigHostedConfigurationVersion67E8BB89": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "SimpleAppConfigApplication04939DED",
        },
        "ConfigurationProfileId": Object {
          "Ref": "SimpleAppConfigConfigurationProfile78536DC3",
        },
        "Content": "{\\"StringKey\\":\\"StringValue\\",\\"BooleanKey\\":true,\\"NumberKey\\":123,\\"CompositeKey\\":{\\"StringKey\\":\\"StringValue\\",\\"BooleanKey\\":false,\\"NumberKey\\":456},\\"ArrayKey\\":[{\\"StringKey\\":\\"StringValue1\\",\\"BooleanKey\\":true,\\"NumberKey\\":123},{\\"StringKey\\":\\"StringValue2\\",\\"BooleanKey\\":false,\\"NumberKey\\":456}]}",
        "ContentType": "application/json",
      },
      "Type": "AWS::AppConfig::HostedConfigurationVersion",
    },
    "SimpleAppConfigImmediateDeployStrategy781EA584": Object {
      "Properties": Object {
        "DeploymentDurationInMinutes": 0,
        "FinalBakeTimeInMinutes": 0,
        "GrowthFactor": 100,
        "Name": "ImmediateDeployment",
        "ReplicateTo": "NONE",
      },
      "Type": "AWS::AppConfig::DeploymentStrategy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
